packages/medusa/src/services/product.ts

```diff
@@ -0,0 +1,1067 @@
+      return [[], false]
+    }
+    const PRICES_RELATION = "variants.prices"
+
+    return omitRelationIfExists(relations, PRICES_RELATION)
+  }
+
+  /**
+   * Lists products based on the provided parameters and includes the count of
+   * products that match the query.
+   * @param selector - an object that defines rules to filter products
+   *   by
+   * @param config - object that defines the scope for what should be
+   *   returned
+   * @return an array containing the products as
+   *   the first element and the total count of products that matches the query
+   *   as the second element.
+   */
+  async listAndCount(
+    selector: FilterableProductProps | Selector<Product>,
+    config: FindProductConfig = {
+      relations: [],
+      skip: 0,
+      take: 20,
+      include_discount_prices: false,
+    }
+  ): Promise<[Product[], number]> {
+    return await this.atomicPhase_(async (manager) => {
+      const productRepo = manager.getCustomRepository(this.productRepository_)
+
+      const { q, query, relations } = this.prepareListQuery_(selector, config)
+
+      const [filteredRelations, shouldLoadStrategyPrices] =
+        this.omitPriceRelationIfExists_(relations)
+
+      let products
+      let count
+      if (q) {
+        ;[products, count] = await productRepo.getFreeTextSearchResultsAndCount(
+          q,
+          query,
+          filteredRelations
+        )
+      } else {
+        ;[products, count] = await productRepo.findWithRelationsAndCount(
+          filteredRelations,
+          query
+        )
+      }
+
+      if (shouldLoadStrategyPrices) {
+        const productsWithAdditionalPrices = await this.setAdditionalPrices(
+          products,
+          {
+            cart_id: config.cart_id,
+            currency_code: config.currency_code,
+            customer_id: config.customer_id,
+            include_discount_prices: config.include_discount_prices,
+            region_id: config.region_id,
+          }
+        )
+
+        return [productsWithAdditionalPrices, count]
+      } else {
+        return [products, count]
+      }
+    })
+  }
+
+  /**
+   * Return the total number of documents in database
+   * @param {object} selector - the selector to choose products by
+   * @return {Promise} the result of the count operation
+   */
+  async count(selector: Selector<Product> = {}): Promise<number> {
+    return await this.atomicPhase_(async (manager) => {
+      const productRepo = manager.getCustomRepository(this.productRepository_)
+      /** revisit */
+      const query = buildQuery(selector)
+      return await productRepo.count(query)
+    })
+  }
+
+  /**
+   * Gets a product by id.
+   * Throws in case of DB Error and if product was not found.
+   * @param productId - id of the product to get.
+   * @param config - object that defines what should be included in the
+   *   query response
+   * @return the result of the find one operation.
+   */
+  async retrieve(
```

Comment:
suggestion: For the 3 different retrieve method, an improvement can be done, I suggest the following (which is a common approach) and would like to here your opinion on that:
```ts
private async retrieve_(
  selector: Selector<Product>,
  config: FindProductConfig = {
    include_discount_prices: false,
  }
) {
  const transactionManager = this.transactionManager_ ?? this.manager_

  const productRepo = transactionManager.getCustomRepository(this.productRepository_)

  const { relations, ...query } = buildQuery(selector, config)

  const [filteredRelations, shouldLoadStrategyPrices] =
    this.omitPriceRelationIfExists_(relations)

  const product = await productRepo.findOneWithRelations(
    filteredRelations,
    query
  )

  if (!product) {
    const selectorConstraints = Object
      .entries(selector)
      .map(([key, value] => `${key}: ${value}`)).join(', ')
    throw new MedusaError(
      MedusaError.Types.NOT_FOUND,
      `Product with ${selectorConstraints} was not found`
    )
  }

  return shouldLoadStrategyPrices
    ? await this.setAdditionalPrices(product, {
        cart_id: config.cart_id,
        currency_code: config.currency_code,
        customer_id: config.customer_id,
        include_discount_prices: config.include_discount_prices,
        region_id: config.region_id,
      })
    : product
}

async retrieve(
  productId: string,
  config: FindProductConfig = {
    include_discount_prices: false,
  }
): Promise<Product> {
  return await this.atomicPhase_(async (manager) => {
    return await this.retrieve_({ id: productId }, config)
  })
}

async retrieveByHandle(
  productHandle: string,
  config: FindProductConfig = {}
): Promise<Product> {
  return await this.atomicPhase_(async (manager) => {
    return await this.retrieve_({ handle: productHandle }, config)
  })
}

async retrieveByExternalId(
  externalId: string,
  config: FindProductConfig = {}
): Promise<Product> {
  return await this.atomicPhase_(async (manager) => {
    return await this.retrieve_({ externa_id: externalId }, config)
  })
}
```

Reply:
This is super nice! Great suggestion @adrien2p 😍